### PR TITLE
Integrate v7 actors

### PR DESCRIPTION
**Pull request by arajasek**
Draft skeleton, basically does all the v7 integration except for FFI call for verification.

Blocked by actors work getting in (https://github.com/filecoin-project/specs-actors/pull/1529)

### CODE REVIEW

**Comment by arajasek** [comment]
Needs to be changed to a `next` or some such branch.

Or we can just land in master with an upgrade epoch set to after the rapture.

File: extern/sector-storage/ffiwrapper/verifier_cgo.go [hunk 1]
```diff
@@ -120,6 +122,11 @@ func (proofVerifier) VerifyAggregateSeals(aggregate proof5.AggregateSealVerifyPr
 	return ffi.VerifyAggregateSeals(aggregate)
 }
 
+func (proofVerifier) VerifyReplicaUpdate(update proof7.ReplicaUpdateInfo) (bool, error) {
+	v := ffi.FunctionsSectorUpdate{}
```
Comment (arajasek) [inline extern/sector-storage/ffiwrapper/verifier_cgo.go:126]: I see conversation about this [here](https://github.com/filecoin-project/filecoin-ffi/pull/208#issuecomment-966457279), but I'm not sure why this breaks the pattern (why isn't the verify call in proofs.go like the others).

@ZenGround0 maybe you know?

File: chain/consensus/filcns/upgrades.go [hunk 2]
```diff
@@ -1170,7 +1186,97 @@ func upgradeActorsV6Common(
...
+func PreUpgradeActorsV7(ctx context.Context, sm *stmgr.StateManager, cache stmgr.MigrationCache, root cid.Cid, epoch abi.ChainEpoch, ts *types.TipSet) error {
+	// Use half the CPUs for pre-migration, but leave at least 3.
+	workerCount := runtime.NumCPU()
+	if workerCount <= 4 {
+		workerCount = 1
+	} else {
+		workerCount /= 2
+	}
+
+	//TODO: nv15
```
Comment (magik6k) [inline chain/consensus/filcns/upgrades.go:1247]: Followup PR?

File: chain/vm/gas_v0.go [hunk 3]
```diff
@@ -228,8 +227,14 @@ func (pl *pricelistV0) OnVerifyAggregateSeals(aggregate proof5.AggregateSealVeri
 	return newGasCharge("OnVerifyAggregateSeals", perProof*num+step.Lookup(num), 0)
 }
 
+// OnVerifyReplicaUpdate
+func (pl *pricelistV0) OnVerifyReplicaUpdate(update proof7.ReplicaUpdateInfo) GasCharge {
+	// TODO: do the thing
```
Comment (magik6k) [inline chain/vm/gas_v0.go:232]: Guessing also a followup PR?

File: extern/sector-storage/ffiwrapper/verifier_cgo.go (hunk 1 above)
Comment (magik6k) [reply extern/sector-storage/ffiwrapper/verifier_cgo.go:126]: Yeah, we need to add those, see https://github.com/filecoin-project/filecoin-ffi/blob/master/proofs.go#L27 for example. (we want it there, so non-lotus things can use go types more easily without importing lotus)

File: extern/sector-storage/mock/mock.go [hunk 4]
```diff
@@ -547,6 +549,11 @@ func (m mockVerifProver) VerifyAggregateSeals(aggregate proof5.AggregateSealVeri
 	return ok, nil
 }
 
+// TODO: do the thing
+func (m mockVerifProver) VerifyReplicaUpdate(update proof7.ReplicaUpdateInfo) (bool, error) {
+	return false, nil
```
Comment (magik6k) [inline extern/sector-storage/mock/mock.go:554]: `return true, nil` should be reasonable for testing, we can do fancier things later

File: itests/ccupgrade_test.go [hunk 5]
```diff
@@ -32,7 +32,7 @@ func runTestCCUpgrade(t *testing.T, upgradeHeight abi.ChainEpoch) {
 	ctx := context.Background()
 	blockTime := 5 * time.Millisecond
 
-	client, miner, ens := kit.EnsembleMinimal(t, kit.MockProofs(), kit.LatestActorsAt(upgradeHeight))
+	client, miner, ens := kit.EnsembleMinimal(t, kit.MockProofs(), kit.TurboUpgradeAt(upgradeHeight))
```
Comment (magik6k) [inline itests/ccupgrade_test.go:35]: (We'll probably end up transforming ccupgrade tests into snapdeals tests, but for now this is ok)

**Review by magik6k — COMMENTED** [review]
Mostly some notes/nits

Also, wdpost test is unhappy (doesn't seem to be flaky, restarting didn't help)

File: chain/vm/gas_v0.go (hunk 3 above)
Comment (arajasek) [reply chain/vm/gas_v0.go:232]: Yes.

File: chain/consensus/filcns/upgrades.go (hunk 2 above)
Comment (arajasek) [reply chain/consensus/filcns/upgrades.go:1247]: This is actually ready, will do.

File: extern/sector-storage/mock/mock.go (hunk 4 above)
Comment (arajasek) [reply extern/sector-storage/mock/mock.go:554]: Ack

File: itests/ccupgrade_test.go (hunk 5 above)
Comment (arajasek) [reply itests/ccupgrade_test.go:35]: putting in a note at top of the file

**Comment by arajasek** [comment]
This is now complete, good to merge after review

File: chain/vm/runtime.go [hunk 6]
```diff
@@ -142,7 +142,7 @@ func (rt *Runtime) StorePut(x cbor.Marshaler) cid.Cid {
 
 var _ rt0.Runtime = (*Runtime)(nil)
 var _ rt5.Runtime = (*Runtime)(nil)
-var _ rt6.Runtime = (*Runtime)(nil)
```
Comment (ZenGround0) [inline chain/vm/runtime.go:145]: nit why do you need to keep around assert on rt5 but not rt6?  Can you remove all but rt7? If there's a breaking change to the runtime interface, same method name different parameter types this pattern won't work.

File: chain/vm/syscalls.go [hunk 7]
```diff
@@ -286,13 +288,27 @@ func (ss *syscallShim) VerifyAggregateSeals(aggregate proof5.AggregateSealVerify
...
+
 	if !ok {
 		return fmt.Errorf("invalid aggregate proof")
 	}
 
 	return nil
 }
 
+func (ss *syscallShim) VerifyReplicaUpdate(update proof7.ReplicaUpdateInfo) error {
+	ok, err := ss.verifier.VerifyReplicaUpdate(update)
```
Comment (ZenGround0) [inline chain/vm/syscalls.go:300]: With new ffi imported does this PR already give lotus correct snap deals update proof verification?  Or is there another PR to follow landing this all the way?

File: itests/wdpost_test.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ import (
 )
 
 func TestWindowedPost(t *testing.T) {
-	kit.Expensive(t)
+	//kit.Expensive(t)
```
Comment (ZenGround0) [inline itests/wdpost_test.go:26]: What's this about? Should we delete this line?

**Review by ZenGround0 — APPROVED** [review]

File: itests/wdpost_test.go (hunk 8 above)
Comment (arajasek) [reply itests/wdpost_test.go:26]: Oops, that was me debugging test failure thanks for busting me.

File: chain/actors/builtin/miner/v7.go [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+		SealProof:             v7.SealProof,
+		SealedCID:             v7.SealedCID,
+		DealIDs:               v7.DealIDs,
+		Activation:            v7.Activation,
+		Expiration:            v7.Expiration,
+		DealWeight:            v7.DealWeight,
+		VerifiedDealWeight:    v7.VerifiedDealWeight,
+		InitialPledge:         v7.InitialPledge,
+		ExpectedDayReward:     v7.ExpectedDayReward,
+		ExpectedStoragePledge: v7.ExpectedStoragePledge,
```
[the remark this answers is not in the document]
Comment (arajasek) [reply chain/actors/builtin/miner/v7.go:551]: Absolutely yes.

File: chain/vm/runtime.go (hunk 6 above)
Comment (arajasek) [reply chain/vm/runtime.go:145]: Based on a quick offline sync, we think it's best to just to have the assertion for every actors version.

If the breaking change @ZenGround0 describes above does happen, we will have a broader refactor to make here.

File: chain/vm/syscalls.go (hunk 7 above)
Comment (arajasek) [reply chain/vm/syscalls.go:300]: gives it (hopefully) correct verification!

**Pull request merged**